src/gtk/toga_gtk/widgets/detailedlist.py

```diff
@@ -1,92 +1,231 @@
-        self.native.add(self.treeview)
-        self.native.set_min_content_width(self.interface.MIN_WIDTH)
-        self.native.set_min_content_height(self.interface.MIN_HEIGHT)
 
-    def change_source(self, source):
-        self.treeview.set_model(None)
-        self.store.change_source(source)
-        for i, node in enumerate(source):
-            self.insert(i, node)
+        self.gtk_on_select_signal_handler = self.list_box.connect(
+            'row-selected', self.gtk_on_row_selected)
+
+        self.right_click_gesture = Gtk.GestureMultiPress.new(self.list_box)
+        self.right_click_gesture.set_button(3)
+        self.right_click_gesture.set_propagation_phase(Gtk.PropagationPhase.BUBBLE)
+        self.right_click_gesture.connect("pressed", self.gtk_on_right_click)
+
+    def row_factory(self, item):
+        """
+        Args:
+            item (:obj:`Row`)
+        Returns:
+            Returns a (:obj:`TextIconRow`)
+        """
+        new_item = TextIconRow(self.interface.factory, self, item)
+        item._impl = new_item
+        return new_item
+
+    def destroy(self):
+        self.disconnect(self.gtk_on_select_signal_handler)
+        super().destroy()
 
-        self.treeview.set_model(self.store)
+    def change_source(self, source):
+        """
+        Args:
+            source (:obj:`ListSource`)
+        """
+        self.store.remove_all()
+        for item in source:
+            self.store.append(
+                self.row_factory(item))
+
+        # We can't know the dimensions of each row (and thus of the list) until gtk allocates
+        # space for it. Gtk does emit `size-allocate` after allocation but I couldn't find any
+        # guarantees that the rows have their sizes allocated in the order they are inserted
+        # in the `ListStore` and in my opinion that's unlikely to be the case.
+
+        # Therefore we would need to wait for `size-allocate` on all rows and either update
+        # the visibility of the buttons on all `size-allocates` or have a counter and only do
+        # it on the last `size-allocate`. Obviously none of those options are desirable.
+
+        # Fortunately functions added with `idle_add` are run when gtk is idle and thus after
+        # any size allocation. This solves our problem and from the perspective of the user
+        # happens immediately.
+
+        # Even though we are adding the callback to the global loop, it only runs once.
+        # This is what the lambda is for. If a callback returns `False` then it's not ran again.
+        # I used a lambda because returning `False` from `self._list_items_changed()` would mean
+        # returning `False` on success.
+        GLib.idle_add(lambda: not self._list_items_changed())
 
     def insert(self, index, item):
-        self.store.insert(item)
+        """
+        Args:
+            index (int)
+            item (:obj:`Row`)
+        """
+        new_item = self.row_factory(item)
+
+        if index == -1:
```

Comment:
Why is this special case required?

Reply:
That was a mistake. In Toga -1 means last item but on Gtk it means not present. I thought -1 meant it was the last item in the list and that `insert` didn't accept -1, thus I used `append`.
Calling `row_factory()` before `item._impl.get_index()` was replacing `item._impl` with a new `Gtk.ListBoxRow` which of course was not yet in the list. I fixed it and changed some variables names' to avoid future confusion.

Also changed `STYLES`to `TOGA_DEFAULT_STYLES`.

Reply:
Hmm... looking at [the docs](https://web.archive.org/web/20210410062839/https://developer.gnome.org/gtk3/stable/GtkListStore.html#gtk-list-store-insert) -1 should be a valid index for `insert()`:

>  If position is -1 or is larger than the number of rows on the list, then the new row will be appended to the list.

But `insert(-1, item)` gives: `OverflowError: -1 not in range 0 to 4294967295`. This might be a bug on Gtk, but in any case I also made the mistake of calling `row_factory()` before `item._impl.get_index()`.

I'm using archive.org because the live page is currently giving me a 404. Since I'm on Debian stable the archive should be recent enough.

Reply:
Oh I was reading the docs for `Gtk.ListStore` not `Gio.ListStore`.